src/include/migraphx/rewrite_reshapes.hpp

```diff
@@ -0,0 +1,206 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#ifndef MIGRAPHX_GUARD_MIGRAPHX_REWRITE_RESHAPES_HPP
+#define MIGRAPHX_GUARD_MIGRAPHX_REWRITE_RESHAPES_HPP
+
+#include <migraphx/config.hpp>
+#include <migraphx/matcher.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/pass_manager.hpp>
+#include <migraphx/common_dims.hpp>
+#include <migraphx/simplify_reshapes.hpp>
+#include <migraphx/eliminate_common_subexpression.hpp>
+#include <migraphx/dead_code_elimination.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+struct rewrite_reshapes_base
+{
+    template <class AxesMap>
+    static instruction_ref insert(module_pass_manager& mpm,
+                                  instruction_ref ins,
+                                  const std::vector<instruction_ref>& inputs,
+                                  const AxesMap&)
+    {
+        return mpm.get_module().insert_instruction(
+            ins, ins->get_operator(), inputs, ins->module_inputs());
+    }
+
+    template <class AxesMap>
+    static bool supports(instruction_ref, std::vector<std::size_t>&, const AxesMap&)
+    {
+        return true;
+    }
+
+    static std::vector<std::size_t> base_dims(instruction_ref ins)
+    {
+        return ins->get_shape().lens();
+    }
+};
+
+template <class T>
+struct rewrite_reshapes
+{
+    std::string name() const { return "rewrite_reshapes"; }
+    struct find_op_reshape_op
+    {
+        std::string op1;
+        std::string op2;
+
+        auto matcher() const
+        {
+            auto reshape =
+                match::name("reshape", "squeeze", "unsqueeze", "flatten")(match::used_once());
+            auto skip_contiguous = [](auto... ms) {
+                return match::arg(0)(match::skip(
+                    match::name("contiguous", "multibroadcast")(match::used_once()))(ms...));
+            };
+            auto pointwise         = match::name(op1)(match::used_once());
+            auto reshape_pointwise = reshape(skip_contiguous(pointwise.bind("x"))).bind("reshape");
+            return match::name(op2)(match::any_of[match::inputs()](reshape_pointwise));
+        }
+
+        template <class F>
+        static instruction_ref find_input_if(instruction_ref start, instruction_ref last, F f)
+        {
+            while(start != last)
+            {
+                if(f(start))
+                    return start;
+                if(start->inputs().size() != 1)
+                    return last;
+                start = start->inputs().front();
+            }
+            return last;
+        }
+
+        static bool match_input(instruction_ref ins, instruction_ref x_ins)
+        {
+            if(ins->inputs().empty())
+                return false;
+            auto input = ins->inputs().front();
+            if(input->name() == "contiguous")
+                return match_input(input, x_ins);
+            return x_ins == input;
```

Comment:
Add simple case where we hit this return.

Reply:
There is coverage for this line. Did you mean the return above?

Reply:
Yes

Reply:
Test added.